net/speedtest/speedtest_server.go

```diff
@@ -0,0 +1,158 @@
+	go (func() {
+		for {
+			conn, err := l.AcceptTCP()
+			if err != nil {
+				// The AcceptTCP will return an error if the listener is closed.
+				return
+			}
+			if numConnections >= maxConnections {
+				continue
+			}
+			connChan <- conn
+		}
+	})()
+
+	for {
+		select {
+		case <-killSignal:
+			return nil
+		case conn := <-connChan:
+			// Handle the connection in a goroutine.
+			go handleConnection(conn, testStateChan)
+			numConnections++
+		case state := <-testStateChan:
+			if state.failed {
+				return state.err
+			} else {
+				// send the results to the results channel to be displayed
+				if resultsChan != nil {
+					resultsChan <- state.results
+				}
+			}
+			numConnections--
+		}
+	}
+}
+
+// handleConnection reads the initial message into a TestConfig struct and
+// determines what test to run. It ignores the config if the type is not
+// download or upload. It sends all errors it comes across as TestStates into
+// the testStateChan channel.
+func handleConnection(conn *net.TCPConn, testStateChan chan TestState) {
+	defer conn.Close()
+	var config TestConfig
+
+	decoder := json.NewDecoder(conn)
+	encoder := json.NewEncoder(conn)
+	err := decoder.Decode(&config)
+
+	// Both return and encode errors that were thrown before the test has started
+	if err != nil {
+		encoder.Encode(TestConfigResponse{Errors: err.Error()})
+		testStateChan <- TestState{failed: true, err: err}
+		return
+	}
+	if config.Version != Version {
+		err = fmt.Errorf("version mismatch! Server is version %d, client is version %d", Version, config.Version)
+		encoder.Encode(TestConfigResponse{Errors: err.Error()})
+		testStateChan <- TestState{failed: true, err: err}
+		return
+	}
+
+	// Start the download test
+	encoder.Encode(TestConfigResponse{Errors: ""})
+	results, err := startTestS(conn, config)
+
+	if err != nil {
+		testStateChan <- TestState{failed: true, err: err, results: nil}
+		return
+	}
+	testStateChan <- TestState{failed: false, err: nil, results: results}
+}
+
+// startTestS runs the server side of the download test. For the given amount of time,
+// it sends randomly generated data in 32 kilobyte blocks. When time's up the function returns
+// and the connection is closed. This function returns an error if the write fails, as well as a
+// slice of results that contains the result of the upload test.
+func startTestS(conn *net.TCPConn, config TestConfig) ([]Result, error) {
+	conn.SetWriteBuffer(LenBufData)
+
+	BufData := make([]byte, LenBufData)
+	sum := 0
+	totalSum := 0
+	lastCalculated := 0.0
+
+	var since time.Duration
+	var results []Result
+
+	for startTime := time.Now(); time.Since(startTime) < config.TestDuration; {
+		// Randomize data
+		_, err := rand.Read(BufData)
+		if err != nil {
+			continue
+		}
+
+		n, err := conn.Write(BufData)
+		if err != nil {
+			// If the write failed, there is most likely something wrong with the connection.
+			return nil, errors.New("server: connection closed unexpectedly")
```

Comment:
Using errors.New returns an undifferentiated error. It is better to wrap the original error with more information:
```
    return nil, fmt.Errorf("server: connection closed unexpectedly: %w", err)
```

That way the caller will get what is still a neterr.connClosed.

In this particular case you could also return err as it is; being a connection closed error seems sufficiently clear.